pkg/envoy/lds/builder.go

```diff
@@ -0,0 +1,599 @@
+					},
+				),
+			},
+		},
+		// Router filter - required to perform HTTP connection management
+		// *Note: the router filter must always be the last filter in the list
+		&xds_hcm.HttpFilter{
+			Name: envoy.HTTPRouterFilterName,
+			ConfigType: &xds_hcm.HttpFilter_TypedConfig{
+				TypedConfig: &any.Any{
+					TypeUrl: envoy.HTTPRouterFilterTypeURL,
+				},
+			},
+		},
+	)
+
+	return hb
+}
+
+// AddFilter adds the given HttpFilter to the builder's filter list.
+// It ensures the HTTP router filter is always the last filter in the list, which
+// is a requirement in Envoy.
+func (hb *httpConnManagerBuilder) AddFilter(filter *xds_hcm.HttpFilter) *httpConnManagerBuilder {
+	if filter == nil {
+		return hb
+	}
+
+	hb.filters = append(hb.filters, filter)
+
+	// If there's only 1 filter, no additional checks required
+	if len(hb.filters) == 1 {
+		return hb
+	}
+
+	routerFilterIndex := -1
+	lastIndex := len(hb.filters) - 1
+
+	for i, f := range hb.filters {
+		if f.Name == envoy.HTTPRouterFilterName {
+			routerFilterIndex = i
+			break
+		}
+	}
+
+	// Guard against the case where multiple router filters are accidentally programmed.
+	// This panics to ensure this never happens in code.
+	if routerFilterIndex != -1 && filter.Name == envoy.HTTPRouterFilterName {
+		panic("multiple router filters not allowed")
+	}
+
+	if routerFilterIndex != lastIndex {
+		routerFilter := hb.filters[routerFilterIndex]
+		hb.filters = append(hb.filters[:routerFilterIndex], hb.filters[routerFilterIndex+1])
+		hb.filters = append(hb.filters, routerFilter)
+	}
+
+	return hb
+}
+
+// LocalReplyConfig sets the given LocalReplyConfig on the builder
+func (hb *httpConnManagerBuilder) LocalReplyConfig(config *xds_hcm.LocalReplyConfig) *httpConnManagerBuilder {
+	hb.localReplyConfig = config
+	return hb
+}
+
+// Tracing sets the Tracing config on the builder
+func (hb *httpConnManagerBuilder) Tracing(config *xds_hcm.HttpConnectionManager_Tracing) *httpConnManagerBuilder {
+	hb.tracing = config
+	return hb
+}
+
+// Build builds the HttpConnectionManager filter from the builder config
+func (hb *httpConnManagerBuilder) Build() (*xds_listener.Filter, error) {
+	connManager := &xds_hcm.HttpConnectionManager{
+		StatPrefix:  hb.statsPrefix,
+		CodecType:   xds_hcm.HttpConnectionManager_AUTO,
+		HttpFilters: hb.filters,
+		RouteSpecifier: &xds_hcm.HttpConnectionManager_Rds{
+			Rds: &xds_hcm.Rds{
+				ConfigSource:    envoy.GetADSConfigSource(),
+				RouteConfigName: hb.routeConfigName,
+			},
+		},
+		AccessLog: envoy.GetAccessLog(),
+		UpgradeConfigs: []*xds_hcm.HttpConnectionManager_UpgradeConfig{
+			{
+				UpgradeType: websocketUpgradeType,
+			},
+		},
+	}
+
+	if hb.tracing != nil {
+		connManager.GenerateRequestId = &wrappers.BoolValue{
+			Value: true,
+		}
+		connManager.Tracing = hb.tracing
+	}
+
+	marshalled, err := anypb.New(connManager)
```

Comment:
Since connManager is built using dynamic configs, we shouldn't use MustMarshalAny, which will panic. We need proto validation to happen for dynamic configs.